diff --git a/video/out/vo_gpu_next.c b/video/out/vo_gpu_next.c
index 0e2e3e0fc4..4d3e869fb2 100644
--- a/video/out/vo_gpu_next.c
+++ b/video/out/vo_gpu_next.c
@@ -1669,9 +1669,9 @@ static int preinit(struct vo *vo)
     mp_mutex_init(&p->dr_lock);
 
     if (gl_opts->shader_cache)
-        cache_init(vo, &p->shader_cache, 10 << 20, gl_opts->shader_cache_dir);
+        cache_init(vo, &p->shader_cache, 150 << 20, gl_opts->shader_cache_dir);
     if (gl_opts->icc_opts->cache)
-        cache_init(vo, &p->icc_cache, 20 << 20, gl_opts->icc_opts->cache_dir);
+        cache_init(vo, &p->icc_cache, 150 << 20, gl_opts->icc_opts->cache_dir);
 
     pl_gpu_set_cache(p->gpu, p->shader_cache.cache);
     p->rr = pl_renderer_create(p->pllog, p->gpu);
