diff --git a/video/out/vo_gpu_next.c b/video/out/vo_gpu_next.c
index 0e2e3e0fc4..c1d7abdd4e 100644
--- a/video/out/vo_gpu_next.c
+++ b/video/out/vo_gpu_next.c
@@ -1669,7 +1669,7 @@ static int preinit(struct vo *vo)
     mp_mutex_init(&p->dr_lock);
 
     if (gl_opts->shader_cache)
-        cache_init(vo, &p->shader_cache, 10 << 20, gl_opts->shader_cache_dir);
+        cache_init(vo, &p->shader_cache, 20 << 20, gl_opts->shader_cache_dir);
     if (gl_opts->icc_opts->cache)
         cache_init(vo, &p->icc_cache, 20 << 20, gl_opts->icc_opts->cache_dir);
 
