diff --git a/meson.build b/meson.build
index 345b591a8c..0e091ca379 100644
--- a/meson.build
+++ b/meson.build
@@ -23,7 +23,7 @@ libavutil = dependency('libavutil', version: '>= 56.70.100')
 libswresample = dependency('libswresample', version: '>= 3.9.100')
 libswscale = dependency('libswscale', version: '>= 5.9.100')
 
-libplacebo = dependency('libplacebo', version: '>=6.338.0')
+libplacebo = dependency('libplacebo', version: '>=6.292.1')
 
 libass = dependency('libass', version: '>= 0.12.2')
 
diff --git a/video/out/mac_common.swift b/video/out/mac_common.swift
index 12d2870add..fd3e2c0018 100644
--- a/video/out/mac_common.swift
+++ b/video/out/mac_common.swift
@@ -111,11 +111,11 @@ class MacCommon: Common {
                                          _ flagsOut: UnsafeMutablePointer<CVOptionFlags>) -> CVReturn
     {
         let frameTimer = mpv?.macOpts.macos_render_timer ?? Int32(RENDER_TIMER_CALLBACK)
-        let signalSwap = { [self] in
-            swapLock.lock()
-            swapTime += 1
-            swapLock.signal()
-            swapLock.unlock()
+        let signalSwap = {
+            self.swapLock.lock()
+            self.swapTime += 1
+            self.swapLock.signal()
+            self.swapLock.unlock()
         }
 
         if frameTimer != RENDER_TIMER_SYSTEM {
