diff --git a/video/out/vo_gpu_next.c b/video/out/vo_gpu_next.c
index 440a324031..f1280036a6 100644
--- a/video/out/vo_gpu_next.c
+++ b/video/out/vo_gpu_next.c
@@ -1693,11 +1693,11 @@ static int preinit(struct vo *vo)
 
     p->shader_cache = pl_cache_create(pl_cache_params(
         .log = p->pllog,
-        .max_total_size  = 10 << 20, // 10 MiB
+        .max_total_size  = 150 << 20, // 150 MiB
     ));
     p->icc_cache = pl_cache_create(pl_cache_params(
         .log = p->pllog,
-        .max_total_size = 10 << 20, // 10 MiB
+        .max_total_size = 150 << 20, // 150 MiB
     ));
     pl_gpu_set_cache(p->gpu, p->shader_cache);
 
